File: cloud/internal/middleware/apikey.go
package middleware

import (
	"context"
	"net/http"
	"time"

	"github.com/kx0101/replayer-cloud/internal/auth"
	"github.com/kx0101/replayer-cloud/internal/store"
)

func APIKeyAuth(s store.Store) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			provided := r.Header.Get("X-API-Key")
			if provided == "" {
				http.Error(w, `{"error":"unauthorized"}`, http.StatusUnauthorized)
				return
			}

			keyHash := auth.HashAPIKey(provided)
			apiKey, err := s.GetAPIKeyByHash(r.Context(), keyHash)
			if err != nil || apiKey == nil {
				http.Error(w, `{"error":"unauthorized"}`, http.StatusUnauthorized)
				return
			}

			if apiKey.ExpiresAt != nil && time.Now().After(*apiKey.ExpiresAt) {
				http.Error(w, `{"error":"api key expired"}`, http.StatusUnauthorized)
				return
			}

			go func() {
				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
				defer cancel()
				_ = s.UpdateAPIKeyLastUsed(ctx, apiKey.ID)
			}()

			ctx := context.WithValue(r.Context(), UserIDContextKey, apiKey.UserID)
			next.ServeHTTP(w, r.WithContext(ctx))
		})
	}
}

File: crates/replayer-cloud/src/middleware/apikey.rs
use axum::http::StatusCode;
use axum::{
    body::Body,
    extract::State,
    http::Request,
    middleware::Next,
    response::{IntoResponse, Response},
};
use uuid::Uuid;

use crate::auth::hash_api_key;
use crate::handler::AppState;

#[derive(Clone, Debug)]
pub struct AuthUserId(pub Uuid);

pub async fn apikey_auth(
    State(state): State<AppState>,
    mut req: Request<Body>,
    next: Next,
) -> Response {
    let provided = req
        .headers()
        .get("X-API-Key")
        .and_then(|v| v.to_str().ok())
        .map(|s| s.to_string());

    let provided = match provided {
        Some(k) if !k.is_empty() => k,
        _ => return (StatusCode::UNAUTHORIZED, r#"{"error":"unauthorized"}"#).into_response(),
    };

    let key_hash = hash_api_key(&provided);
    let api_key_record = match state.store.get_api_key_by_hash(&key_hash).await {
        Ok(Some(k)) => k,
        _ => return (StatusCode::UNAUTHORIZED, r#"{"error":"unauthorized"}"#).into_response(),
    };

    if let Some(expires) = api_key_record.expires_at {
        if chrono::Utc::now() > expires {
            return (StatusCode::UNAUTHORIZED, r#"{"error":"api key expired"}"#).into_response();
        }
    }

    let store = state.store.clone();
    let key_id = api_key_record.id;
    tokio::spawn(async move {
        let _ = store.update_api_key_last_used(key_id).await;
    });

    req.extensions_mut()
        .insert(AuthUserId(api_key_record.user_id));
    next.run(req).await
}
